### ISSUE TITLE
Test redeploy of berkeley testnet (without embedded runtime configs)

### ISSUE THREAD

**Issue by stevenplatt** [MinaProtocol/mina#11138]
This issue is to document testing for deploying the Berkeley testnet without embedded runtime configs. The current deployment embeds the ledger into the docker image - which can cause issue if changing the image within the terraform later, or added new nodes later.

**Comment by stevenplatt** [comment]
- [ ] This item is completed. The original intent of this zenhub issue was to trial what would happen if the existing terraform was modified to extend the network or change nodes without fully tearing down and redeploying. In such a scenario, the new nodes or configuration may not match the accounts existing in the baked mina daemon image that was used in the original deployment. 

A redeploy of the Berkeley network was completed yesterday (Thur June 2), but was done without disabling the embedded runtime config). After reviewing the Terraform configuration and having the additional deploy practice, it is understood that such an extension of an already deployed network is not recommended. This is mostly due to the `null resource` scripts that generate keys during deployment. Terraform does not track a state for `null resource` items, so modification of the network after the initial deployment (even if done by terraform) would not conform or be supportable - since the updated state will lose any association to the items executed by the `null resource` items at the initial deployment. An example of the `null resource` tasks, are the shell scripts that generate the keys and ledger. Using `null resource` is not recommended with Terraform due to the issue of state management. These items are expected to be removed and completed as an external item for the future Mina hard fork deployment. An example of the Terraform `null resource` calling to a shell script can be found here; https://github.com/MinaProtocol/mina/blob/develop/automation/terraform/modules/o1-testnet/scripts.tf

_________________

For documentation, the steps that were required for redeployment of the Berkeley testnet were: 

**Prerequisites:** The nodes keys that are used for deployment are included in the repository for security reasons. These need to first be obtained from another engineer (O(1) does not have a specific location where these are being stored, although they are likely to be stored in a private cloud bucket in the future). Once the key are obtained, they must be placed into the `mina/automation` directory of your local copy of the repository. Terraform will complain about these missing files and refuse to function if the files are not present. 

Within your copy of the new `mina/automation/keys` directory, there must also be the empty files of: 

- [ ] `mina/automation/keys/keysets/berkeley_offline-fish`

- [ ] `mina/automation/keys/keysets/berkeley_offline-whales`

Both of the above files should be empty, and because of this, they may not be included in the copy of the `keys` folder received. If they are not present, they can be created using `touch [FILE_NAME]`. Note that these files don't have any extension name, so creating them usingthe `touch` command within the terminal may be the safest.

With the prerequisite keys information in the cloned repository, and with Google and Kubernetes auhentication and contexts set, the following steps are taken to redeploy the Berkeley testnet:

- [ ]  Modify the Genesis ledger timestamp to the current time. This is needed to force a fresh restart of the network. Otherwise, nodes that were connected to the old deployment, may continue using the old copy of the genesis ledger. 

- [ ] Open a PR and merge the edited ledger into the target branch, in the case of Berkeley testnet, this woul dbe `develop`.

- [ ] Open a PR to then merge the updated `develop` branch into a release branch. In this case, the release branch was `release/2.0.0`. 

- [ ] Once the update is merged into the release branch, manually trigger a CI job which targets the release branch. This is done from the buildkite UI and is required to get a final docker image published, which contains are changes. 

- [ ] Within the testnet Terraform directory on your local machine (`mina/automation/terraform/testnets/berkeley`), issue a `Terraform init` in the terminal. This will sync the terraform state file to your local directory. 

- [ ] Issue a `Terraform Plan` command from the same directory. This will validate you have all kubernetes context and google cloud auth items activated, as well as validate all required files are present, such as keys. If there are errors here, they will need to be resolved before proceeding. 

- [ ] Next issue `Terraform destroy` and respond `yes` to the prompt, this will delete all existing Berkeley testnet items. Wait a few minutes after the success prompt before proceeding - to safely allow the remote cloud provide to replicate the changed resource states. 

- [ ] Once the manually created buildkite CI job completes the steps of building the target Docker image (you can check this by searching for `Berkeley`  in the Buildkite job UI. At this time, there are specific test steps to build a Berkeley-specific image). Copy/paste the name of the resulting docker image tht is built, and update the `mina.tf` within `mina//automation/terraform/testnets/berkeley/` so that `mina_image` and `mina_archive_image` have the updated docker image names. Note that the docker image used for `mina_image` has a Berkeley-specific name, but the image for `mina_archive_image` does not. It will use the standard archive image produced from the CI job. Save the `main.tf` file after the changes. 

- [ ] Issue the `terraform apply` command. If there are no errors, reply `yes` to the terminal prompt to proceed with the changes. This will redeploy the Berkeley testnet. 

**Post Deployment**

The Berekeley nodes are deployed with liveness checks. It will take 10minutes or more for all the nodes to boot properly and reach a green state. Note that items for `db-bootstrap` will appear in an unhealthy state , but this is normal as they are bootstrap jobs that run only once to initialize the Archive databases. 

With everything in a healthy state (except the db-bootstrap items). The following check will confirm everything is healthy post-deployment:

- [ ] Connect to the console of a fish node, check that it has contactd peers, and is syncing. Wait a period of time and confirm the node eventually fully syncs by issuing the `mina client status` command at the node command line.

- [ ] Launch a new node from your local machine (using the updated docker image), and connect it to the testnet. Wait a period of time to confirm that it also can find peers and reach a fully synced state. Checking this status is done the same way, by first connecting to the running docker container, then issuing  the `mina client status` command at the command line. The docker command used to start the container in this case were:
```
docker run -d --name mina \
-p 8302:8302 \
-p 3085:3085 \
minaprotocol/mina-daemon:1.3.1beta1-release-2.0.0-c160fce-bionic-berkeley \
daemon \
--log-level Debug \
--peer-list-url https://storage.googleapis.com/seed-lists/berkeley_seeds.txt
```

- [ ] If the above two items are successful, issue the `mina client status` command a final time at any of the active nodes and take note of the `chain id`. It should have a format similar to `chain id: 9004d8c6b64d867b42f58f8881230882f2a14ecefcdcf40e66df8337a5b70f79`. This id can be compared to the Id from community nodes to confirm they have indeed picked up the new clean ledger and are in the proper synced state. 

**A final note**

When the berkeley testnet was redeployed, the nodes were deployed using the same IP addresses. This is likely because the old IP's had not been released on the cloud provder side, since the network was torn down and redeployed within a few hours window. Because the IP and DNS entries were the same, the DNS entries within `--peer-list-url https://storage.googleapis.com/seed-lists/berkeley_seeds.txt` were still valid and did not need to be replaced. This was confirmed by checking the AWS Route 53 records. If for any reason these entries are no longer valid, it would be required to replace this berkeley_seeds.txt file with a file containing the correct entries.
